zsl/resource/guard.py

```diff
@@ -0,0 +1,358 @@
+            # override GuardedResource.read, but with its security checks
+            def secure_read(self, param, args, data):
+                return specific_resources[param]
+        
+    """
+    method_wrappers = []
+    exception_handlers = [default_error_handler]
+    resource_methods = ['create', 'read', 'update', 'delete']
+
+    def __init__(self, policies=None, method_wrappers=None,
+                 exception_handlers=None):
+        # type: (Optional[List[policies]]) -> None
+
+        self.policies = list(policies) if policies else []
+
+        if method_wrappers:
+            self._method_wrappers = self.method_wrappers + method_wrappers
+        else:
+            self._method_wrappers = list(self.method_wrappers)
+
+        if exception_handlers:
+            self._exception_handlers = self.exception_handlers + \
+                                       exception_handlers
+        else:
+            self._exception_handlers = list(self.exception_handlers)
+
+    def _check_before_policies(self, name, *args, **kwargs):
+        if not any(_before(p, name)(*args, **kwargs) for p in self.policies):
+                raise PolicyViolationError('Policy violation for {} {}'.format(
+                    name, 'before'), code=403)
+
+    def _check_after_policies(self, name, result):
+        if not any(_after(p, name)(result) for p in self.policies):
+                raise PolicyViolationError('Policy violation for {} {}'.format(
+                    name, 'after'), code=403)
+
+    def _wrap(self, method):
+        # type: (Callable) -> Callable
+
+        name = method.__name__
+
+        @wraps(method)
+        def wrapped(*args, **kwargs):
+
+            res = args[0]
+            args = args[1:]
+
+            try:
+                self._check_before_policies(name, *args, **kwargs)
+                rv = _secure_method(res, name)(*args, **kwargs)
+                self._check_after_policies(name, rv)
+
+            except PolicyViolationError as e:
+                rv = self._handle_exception(e, res)
+
+            return rv
+
+        for mw in self._method_wrappers:
+            wrapped = mw(wrapped)
+
+        return wrapped
+
+    def _handle_exception(self, error, resource):
+        rv = None
+        for handler in self._exception_handlers:
+            rv = handler(error, rv, resource)
+
+        return rv
+
+    def __call__(self, cls):
+        if hasattr(cls, '_guard_policies'):
+            self.policies += getattr(cls, '_guard_policies')
+            setattr(cls, '_guard_policies', list(self.policies))
+
+            return cls
+
+        setattr(cls, '_guard_policies', list(self.policies))
+
+        for method_name in self.resource_methods:
+            secure_name = _secure_name(method_name)
+
+            if hasattr(cls, method_name):
+                method = getattr(cls, method_name)
+
+                setattr(cls, method_name, self._wrap(method))
+                setattr(cls, secure_name, method)
```

Comment:
I see the secure_name is the non-secured method (core of the functionality). this is something that got me.